man/man1/wstart.1
.TH WSTART 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wstart \- a program launcher bar for W
.SH SYNOPSIS
.nf
wstart [\fI-f font\fP] [\fI-q\fP] [\fI-t\fP] [\fI-e\fP] [\fI-v\fP|\fI-h\fP] [\fI-l\fP|\fI-c\fP|\fI-r\fP] [\fI"root"\fP]
.fi
.SH DESCRIPTION
\fIWSTart\fP is a W program launcher bar.  From the popup(s) under
the leftmost button(s) you can execute programs that WSTart found in
the given directory.
.PP
As WSTart is intended to be running all the time, it will
re-read directories before opening a popup representing
directory contents.  WSTart will show you only executables and
subdirectories.  If directory contains neither ones, submenu
popup will be canceled!  All files starting with a `.'
character are understood to be either `system' or `hidden' files
and therefore ignored.
.SH OPTIONS
.TP
\fI-f\fP \fIfont name\fP
Changes font used in launch bar and popups.
.TP
.I -t
Remove time show from the launch bar.
.TP
.I -q
Remove quit button from the launch bar.
.TP
.I -e
Report errors (in a dialog).
.TP
.I -h
Extend launch bar horizontally to screen width (default).
.TP
.I -v
Extend launch bar vertically to screen height.
.TP
.I -l
Position launch bar on left / top of the screen (default).
.TP
.I -c
Position launch bar on center of the screen.
.TP
.I -r
Position launch bar on right / bottom of the screen.
.PP
Last argument may be a path which WSTart will use as it's `root'
directory.  If given path doesn't start with `/', it's interpreted as
being relative to the current work directory, which is the default.
.PP
`Root' directory contents are show as buttons in the launch bar.
.SH RELATED FILES
There may be icon and option files associated with programs. WSTart
will check if directory has files with `.icon' and `.options' appended
to the end of the program name.  Directory could have for example
files:
.TP
wterm
executable
.TP
wterm.icon
executable icon
.TP
wterm.options
executable options
.TP
folder.icon
default icon for directories
.TP
file.icon
default icon for files
.PP
If you have the default icons `folder.icon' and `file.icon', they have
to reside in the WSTart `root' directory.
.SH ICON FILES
Icon format is currently PBM (Portable Bitmap, see `pbm' manual).
These can be created with `Icone' icon editor or be converted from
other formats with the free `netpbm' utilities or others (PBM format
is quite well supported).  Popups look best if icons are all the same
size and at most as high as popup text (see `-f' option).
.SH OPTION FILES
Option file is composed of program arguments (and their values)
separated with white space.  Use an empty option file if program
doesn't take any arguments.  If there's no option file WSTart will ask
you for options.  The command line you type can be saved onto disk
from the option dialog.
.PP
Option file rules:
.IP "\-" 8
All characters between values 1 and 32 are regarded as `white space'.
NULL terminates options, so arguments may not contain them.
.IP "\-" 8
If you want to include a string including white space, enclose it
inside single quote pair (for example "-s 'silly example'"). Single
quotes cannot be `quoted'. They will always either start or end a
quote and appear in pairs.
.IP "\-" 8
You can use a directive for inquiring a string from the user.
Directives start with a `%' character and continue to the end of the
line (= are terminated with newline character).  Directives may have
arguments.  Some are optional, but if present, they have to be in the
order described.
.PP
At the moment there's only `%f' directive which will present user with
a fileselector for getting a filename.  The whole directive will be
replaced with the filename user selects. Directive usage is:
.nf
	%f [title] [default mask] [default path] [default file]
.fi
.PP
If user cancels a directive, program won't be run.
.SH OPTION FILE EXAMPLES
Normally option files contain nothing or simple one liners to set
default options to your preferences.  Files can be created just by
using the `Save arguments' option in the options dialog, which will be
presented when program doesn't have an option file.
.nf
	---- an example for wedit -----
	-w 78 -h 16 -t 4 -f atari8.wfnt
	-------------------------------
.fi
.PP
Option files with directive(s) have to be currently created with a
text editor because directives need more than one line.  Below is an
example of a fileselector directive use:
.nf
	------ misc. program ----------
	-width  72
	-height 20
	-file   %f ' Load file: ' *.doc /home/eero/test/doc/ default.doc
	-------------------------------
.fi
.PP
If you would want something more elaborate, feel free to implement
whatever you want to.
.SH INSTALLING
Create a directory hierarchy where you'll put your binaries.
.PP
Move or link the desired binaries into these directories.  You should
have in the main directory only subdirectories and possible a couple
of most important programs and sort rest of the programs into
subdirectories.
.PP
Copy or link desired program icons into same directory with programs
you'll want to associate them with.  Note that you may *not* use
_symbolic_ links with icons as those would have execute bit set and
then WSTart would mistake them for programs.
.PP
Copy or link icons that you want to be defaults to files `folder.icon'
and `file.icon' in the WSTart `root' directory. Program options could
be later created from the WSTart argument dialog.
.PP
Move `wstart' binary into a directory set in your PATH. For example
into the one where you'll keep the W server binary.
.PP
Add WSTart either to `~/.wconfig' or `~/.wrc' W configuration file
with the main directory of your newly created binary directory
hierarchy as an argument.  I myself use:
.nf
	wstart /home/eero/wstart/ &
.fi
.PP
.SH TECHNICAL INFORMATION
Before you'll see a subdirectory popup, following things have
been done:
.IP "\-" 8
Directory contents are read to see how many executables and
directories there are and how much space their names take.
.IP "\-" 8
Menu structure is allocated for them, and the directory entries are
read into it.
.IP "\-" 8
Icons are read for programs which have such.
.IP "\-" 8
Popup is created for the new submenu.
.IP "\-" 8
Popup widget parses menu structure into a graphical representation and
draws it (icons, text and graphics are sent to W server).
.PP
Doing this may take some time if your machine is slow and you
have lots of files in a directory.  Directory contents aren't
sorted in any way.  If you absolutely want programs to appear in
certain order, you might try removing them and then copy / link
them again in that order.  Depending on how your filesystem
stores directory entries this either works or not...  WStart
reads directory entries in the order they are in the directory.
.SH NOTES
WStart children use same stdin/out/err streams as WStart (define
CHILD_REDIRECT in source if you want to disable that).  In case
WStart is run in background without output redirection, WStart
(and programs it runs) ignore SIGGTOU, so that background writes
won't stop them.
.PP
Normally WStart doesn't report errors.  If subdirectory isn't
opened, it either doesn't exist (anymore) or it doesn't contain
directories or executables.  If program doesn't run either it
doesn't exist (anymore) or running it failed somehow (usual
reason being wrong arguments).
.PP
Somebody suggested that WStart should use PATH instead of the
link directory scheme. That's impractical for several reasons:
.IP "\-" 8
There would be *lots* of programs on menus which you'd have no
interest in.
.IP "\-" 8
Menus wouldn't fit on screen nor would programs be logically divided
between them.
.IP "\-" 8
You might not have write rights to the directories, so setting up
option and/or icon files couldn't be done.
.SH BUGS
Probably many, but none that I would know right now.  Option
file parsing would probably be most likely candinate for them.
.PP
On slower (like 8Mhz Atari ST) machines this is unusable.
This depends a bit on how well your disk is cached.
.SH LAST CHANGED
11/1997 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and

wt/wstart/wstart.c
/* 
 * A `Win-95 style' program lauch bar
 * 
 * (w) 1997 by Eero Tamminen
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <dirent.h>
#include <string.h>
#include <limits.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <fcntl.h>
#include <wait.h>
#include <time.h>
#include <Wlib.h>
#include <Wt.h>

/* whether to redirect child standard streams to /dev/null */
#undef CHILD_REDIRECT

/* additional lauch bar element flags */
#define ADD_TIME	1
#define ADD_QUIT	2

/* standard file types / names */
#define ICON_TYPE	".icon"
#define OPTION_TYPE	".options"
#define DIR_ICON	"folder.icon"
#define FILE_ICON	"file.icon"

#define WIN_BORDER	4		/* width of a W window border */

/* enough space for all pathnames available on this system */
#define MAX_PATH_LEN		(_POSIX_PATH_MAX + _POSIX_NAME_MAX)

/* default icons set in initialize() */
static BITMAP *FolderIcon;
static BITMAP *FileIcon;

static void
directory_cb(widget_t *w, wt_menu_item *menu);

/* 
 * utility functions
 */

static int ReportErrors;

static void
report_error(const char *error)
{
	if (!ReportErrors) {
		return;
	}
	wt_dialog(NULL, error, WT_DIAL_ERROR, " WStart error ", "Cancel", NULL);
}

/* read submenu directory from menu end marker and appen given filename to
 * that.  Return results in a buffer large enough for longest filename OS
 * can have.
 */
static char *
menu_path(const wt_menu_item *item)
{
	/* Here's a temporary buffer for storing file paths and names.  It
	 * has to be large enough for the longest possible file path + name
	 * string.  Functions using this are ATM initialize(), run_cb(),
	 * directory_cb() and their subfunctions.
	 */
	static char buffer[MAX_PATH_LEN];
	const char *name;

	if (!item) {
		/* just return enough space for filepath+name */
		return buffer;
	}
	name = item->string;

	/* search last */
	while (item->type != MenuEnd) {
		item++;
	}
	/* new file path+name to buffer */
	strcpy(buffer, item->string);
	strcpy(&buffer[strlen(buffer)], name);
	return buffer;
}

static inline void
free_path(char *path)
{
	/* is static one from above, doesn't need to free */
}

/* 
 * program execution
 */

/* count arguments in the string and separate them with zeros */
static int
separate_arguments(char *args)
{
	int count;

	/* zero and skip white space */
	while (*args && *args <= ' ') {
		*(args++) = '\0';
	}
	for (count = 0; *args; count++) {
		switch (*args) {
			case '%':
				/* special directive to end of line */
				while (*args && *args != '\n') {
					args++;
				}
				break;
			case '\'':
				/* skip white space enclosed... */
				*args++ = '\0';
				do {
					/* ...quoted argument */
					while (*args && *args != '\'') {
						args++;
					}
				} while (*args && *(args+1) > ' ');
				if (*args == '\'') {
					*args++ = '\0';
				}
				break;
			default:
				/* skip argument */
				while (*args > ' ') {
					args++;
				}
		}
		/* zero and skip white space */
		while (*args && *args <= ' ') {
			*(args++) = '\0';
		}
	}
	return count;
}

/* allocate and return either NULL or string from user (ATM there's
 * only `%f' directive for getting a filename with a fileselector).
 */
static char *
alloc_directive(char *cmd)
{
#define DIRECTIVE_ARGS 4
	const char *opt[DIRECTIVE_ARGS];
	char *arg, *options;
	int idx, count;

	/* not `%f' directive? */
	if (cmd[0] != '%' || cmd[1] != 'f' || cmd[2] > ' ' ) {
		report_error("Unknown option directive!");
		return NULL;
	}
	options = cmd + 2;

	/* get directive arguments */
	count = separate_arguments(options);
	if (count > DIRECTIVE_ARGS) {
		count = DIRECTIVE_ARGS;
	}
	for (idx = 0; idx < count; idx++) {
		while(!*options) {
			options++;
		}
		opt[idx] = options;
		options += strlen(options);
	}
	for (idx = count; idx < DIRECTIVE_ARGS; idx++) {
		opt[idx] = "";
	}

	/* here could be other directives too... */
	arg = wt_fileselect(NULL, opt[0], opt[2], opt[1], opt[3]);
	return arg;
#undef DIRECTIVE_ARGS
}

/* get program arguments either from file or user and return the string */
static char *
get_arguments(const char *name, const char *file)
{
#define ARG_LEN 128
	struct stat st;
	char *args;
	FILE *fp;

	if ((fp = fopen(file, "r"))) {
		stat(file, &st);
		if (!st.st_size) {
			fclose(fp);
			return NULL;
		}
		if (!(args = malloc(st.st_size+1))) {
			fclose(fp);
			report_error("Options file allocation failed!");
			return NULL;
		}
		if (fread(args, st.st_size, 1, fp) != 1) {
			fclose(fp);
			free(args);
			report_error("Options file read failed!");
			return NULL;
		}
		args[st.st_size] = '\0';
		fclose(fp);
	} else {
		/* ask arguments from the user */
		if (!(args = malloc(ARG_LEN+1))) {
			report_error("Argument allocation failed!");
			return NULL;
		}
		args[0] = 0;
		for(;;) {
			switch(wt_entrybox(NULL, args, ARG_LEN, name, "arguments:",
			"Execute program", "Save arguments", NULL)) {
				case 1:
					return args;
				case 2:
					if ((fp = fopen(file, "w"))) {
						fwrite(args, strlen(args), 1, fp);
						fclose(fp);
					}
					break;
				default:
					free(args);
					return NULL;
			}
		}
	}
	return args;
#undef ARG_LEN
}

static inline void
free_arguments(char *args)
{
	free(args);
}

/* Read the program options file if available and allocate / compose an
 * argument string array.
 */
static char **
alloc_argv(const char *name, const char *file)
{
	char *strings, *args, **argv;
	int count, idx, len;

	if(!(strings = get_arguments(name, file))) {
		return NULL;
	}
	/* program name + separate entries + NULL */
        count = 1 + separate_arguments(strings);
	if (!(argv = malloc(sizeof(char*) * (count+1)))) {
		free_arguments(strings);
		report_error("Argv[] allocation failed!");
		return NULL;
	}
	/* setup argument pointers */
	args = strings;
	argv[0] = strdup(name);
	for (idx = 1; idx < count; idx++) {
		while(!*args) {
			args++;
		}
		len = strlen(args);
		if (*args == '%') {
			/* subdivides the argument string and asks user */
			argv[idx] = alloc_directive(args);
			if (!argv[idx]) {
				/* user canceled input */
				while (--idx >= 0) {
					free(argv[idx]);
				}
				free(argv);
				free_arguments(strings);
				return NULL;
			}
		} else {
			argv[idx] = strdup(args);
		}
		args += len;
	}
	argv[idx] = NULL;
	free_arguments(strings);
	return argv;
}

static void
free_argv(char **argv)
{
	int idx;
	for (idx = 0; argv[idx]; idx++) {
		free(argv[idx]);
	}
	free(argv);
}


static void
run_cb(widget_t *w, wt_menu_item *item)
{
	char *program, **argv;
	struct stat st;
	int len;
#ifdef CHILD_REDIRECT
	int fh;
#endif
	/* ATM static & same on all calls! */
	program = menu_path(item);

	len = strlen(program);
	/* add option type temporarily */
	strcpy(&program[len], OPTION_TYPE);
	/*
	 * This has to be done before fork.  Forked processes share file
	 * handles (W socket) and Wt opens connection only once.
	 */
	argv = alloc_argv(item->string, program);
	program[len] = '\0';
	if (!argv) {
		return;
	}
	if (stat(program, &st) < 0) {
		report_error("Program unaccessible!");
	} else {
		if (!fork()) {
			w_exit();
#ifdef CHILD_REDIRECT
			/* above should close W socket and stuff below
			 * remaps standard streams. Other file descriptors
			 * shouldn't be open.
			 */
			if ((fh = open("/dev/null", O_RDWR)) >= 0) {
				dup2(fh, 0);
				dup2(fh, 1);
				dup2(fh, 2);
				close(fh);
			} else {
				fprintf(stderr, "WStart: can't open /dev/null\n");
			}
#endif
			execvp(program, argv);
			exit(-999);
		}
	}
	free_argv(argv);
	free_path(program);
}


/* 
 * subdirectory handling
 *
 * The directory path of which contents menu represents is stored in the
 * `string' member of the last menu item (end marker).
 *
 * Stuff that menu has pointers into, is allocated in one block except
 * for the possible induvidual program icons (default icons are static).
 */

enum { ACCEPT_FILE = 1, ACCEPT_DIR };

/* 0: file is NOT runnable by given user
 * 1: normal file
 * 2: directory
 */
static int
accept(const char *name, const uid_t uid, const gid_t gid)
{
	struct stat st;
	int mode;

	if (stat(name, &st) < 0)
		return 0;

	if (S_ISDIR(st.st_mode)) {
		mode = ACCEPT_DIR;
	} else {
		mode = ACCEPT_FILE;
	}
	/* executable? */
	if (st.st_mode & S_IXOTH)
		return mode;
	if ((st.st_mode & S_IXGRP) && st.st_gid == gid)
		return mode;
	if ((st.st_mode & S_IXUSR) && st.st_uid == uid)
		return mode;

	return 0;
}

/* Read directory, count accepted menu entries and then compose menu
 * structure of them.
 *
 * String buffer pointed to by `path' is used in this function so it should
 * be long enough for the longest possible file path + name!
 *
 * Returns NULL or an error message.
 */
static wt_menu_item *
read_menu(char *path)
{
	int mode, index, count, pathlen, namelen, len;
	wt_menu_item *menu, *item;
	struct dirent *entry;
	char *names;
	uid_t uid;
	gid_t gid;
	DIR *dfd;

	if(!(dfd = opendir(path))) {
		report_error("Directory open failed!");
		return NULL;
	}
	gid = getgid();
	uid = getuid();
	pathlen = strlen(path);
	count = namelen = 0;
	/* 
	 * files starting with `.'  are links to current and parent
	 * directory or configuration files / directories so they are
	 * ignored too.
	 */
	while((entry = readdir(dfd))) {
		if (entry->d_name[0] == '.') {
			continue;
		}
		strcpy(&path[pathlen], entry->d_name);
		if ((mode = accept(path, uid, gid))) {
			count++;
			namelen += strlen(entry->d_name) + 1;
			if (mode == ACCEPT_DIR)
				namelen++;
		}
	}
	path[pathlen] = '\0';
	if (!count) {
		closedir(dfd);
		report_error("No usable entries!");
		return NULL;		/* no entries */
	}
	count++;
	/* menu items, file names, directory path */
	menu = malloc(sizeof(wt_menu_item)*count + namelen + strlen(path)+2);
	if (!menu) {
		closedir(dfd);
		report_error("Directory allocation failed!");
		return NULL;
	}
	names = (char *)(menu + count);
	memset(menu, 0, names - (char *)menu);
	count--;

	index = 0;
	rewinddir(dfd);
	/* re-read directory entries into the menu */
	while(index < count && (entry = readdir(dfd))) {
		if (entry->d_name[0] == '.') {
			continue;
		}
		strcpy(&path[pathlen], entry->d_name);
		if ((mode = accept(path, uid, gid))) {

			/* check that name fits to allocation in case dir
			 * contents are changed in the meanwhile...
			 */
			len = strlen(entry->d_name) + 1;
			if (namelen < len) {
				break;
			}
			item = &(menu[index++]);
			strcpy(names, entry->d_name);
			item->string = names;
			namelen -= len;
			names += len;

			if (mode == ACCEPT_DIR) {
				namelen--;
				*(names-1) = '/';
				*(names++) = '\0';
				item->select_cb = directory_cb;
				item->type = MenuSub;
			} else {
				item->select_cb = run_cb;
				item->type = MenuItem;
			}
		}
	}
	path[pathlen] = '\0';
	closedir(dfd);

	strcpy(names, path);
	/* store `menu' path to the end mark item */
	menu[index].string = names;
	menu[index].type = MenuEnd;
	return menu;
}

static inline void
free_menu(wt_menu_item *data)
{
	free(data);
}

/* 
 * load icons for all the menu items
 *
 * String buffer pointed to by `path' is used in this function so it should
 * be long enough for the longest possible file path + name!
 */
static void
load_icons(char *path, wt_menu_item *item)
{
	int len = strlen(path);
	while (item->type != MenuEnd) {
		if (item->type == MenuItem) {
			strcpy(&path[len], item->string);
			strcpy(&path[len + strlen(item->string)], ICON_TYPE);
			item->icon = w_readpbm(path);
			if (!item->icon) {
				item->icon = FileIcon;
			}
		} else {
			/* sub-directory */
			item->icon = FolderIcon;
		}
		item++;
	}
	path[len] = '\0';
}

static void
release_icons(wt_menu_item *item)
{
	while (item->type != MenuEnd) {
		if (item->icon != FolderIcon && item->icon != FileIcon) {
			/* icon data should be on same block */
			w_freebm(item->icon);
		}
		item++;
	}
}

/*
 * Release given old submenu and compose a new one from the current
 * contents of the pointed directory or return an error message.
 * Reads also icons for the newly created submenu items.
 */
static void
directory_cb(widget_t *w, wt_menu_item *item)
{
	char *buffer;

	/* free old submenu and it's contents */
	if (item->sub) {
		release_icons(item->sub);
		/* rest is allocated in one piece in read_menu() */
		free_menu(item->sub);
		item->sub = NULL;
	}

	/* ATM static & same on all calls! */
	buffer = menu_path(item);

	/* modifies buffer, but returns it back to what it was */
	if (!(item->sub = read_menu(buffer))) {
		free_path(buffer);
		return;
	}

	/* modifies buffer, but returns it back to what it was */
	load_icons(buffer, item->sub);
	free_path(buffer);
}

/* 
 * other functionality
 */

static void
timer_cb(const long arg)
{
	widget_t *wp = (widget_t *)arg;
	char label[6];
	time_t value;
	struct tm *t;

	value = time(&value);
	t = localtime(&value);
	sprintf(label, "%02d:%02d", t->tm_hour, t->tm_min);
	wt_setopt(wp, WT_LABEL, label, WT_EOL);
	/* update once every minute or so */
	wt_addtimeout(59000, timer_cb, arg);
}

static void
draw_icon_cb(widget_t *w, short x, short y, BITMAP *bm)
{
	WWIN *win = wt_widget2win(w);
	w_putblock(bm, win, x, y);
}

static void
quit_cb(widget_t *w, int pressed)
{
	if (!pressed)
		wt_break(1);
}

/* remove zombies:  ask for information about terminated children, until
 * there is no more, just for the case that several processes died at the
 * same time. Posix systems should do this automatically when SIG_CHLD
 * is ignored.
 */
#if !defined(POSIX) && !defined(SysV) && !defined(linux)
static void
sigchild(int signal)
{
  int status;
  while (waitpid(-1, &status, WNOHANG) > 0);
}
#endif

/* 
 * Initialization
 */

/* parse command line argument(s), compose `menu'[2] structures with program
 * root directory set to the last item in the menu, read default icons.
 */
static wt_menu_item *
initialize(int argc, char *argv[],
	   int *elements, int *orient, int *align,
	   const char **font, const char **error)
{
	static wt_menu_item menu[2];	/* static -> initialized to zero */
	int len, idx;
	char *buffer;
	DIR *dfd;

#if defined(POSIX) || defined(SysV) || defined(linux)
	/* Posix */
	signal(SIGCHLD, SIG_IGN);
#else
	/* BSDish OS needs signal handler for getting rid of zombie
	 * children.
	 */
	signal(SIGCHLD, sigchild);
#endif
#ifdef SIGTTOU
	/* Allow background writes (from programs) to control
	 * terminal by ignoring terminal I/O signals.
	 */
	signal(SIGTTOU, SIG_IGN);
#endif

	/* default options */
	*align = AlignTop;
	*orient = OrientHorz;
	*font = *error = NULL;
	*elements = ADD_TIME | ADD_QUIT;

	idx = 0;
	while(++idx < argc && argv[idx][0] == '-') {
		switch(argv[idx][1]) {
			case 'e':
				ReportErrors ^= 1;
				break;
			case 't':
				*elements ^= ~ADD_TIME;
				break;
			case 'q':
				*elements ^= ~ADD_QUIT;
				break;
			case 'h':
				*orient = OrientHorz;
				break;
			case 'v':
				*orient = OrientVert;
				break;
			case 'l':
				*align = AlignTop;
				break;
			case 'c':
				*align = AlignCenter;
				break;
			case 'r':
				*align = AlignBottom;
				break;
			case 'f':
				if (++idx < argc) {
					*font = argv[idx];
				} else {
					*error = "font name missing";
					return NULL;
				}
				break;
			default:
				*error = "unrecognized option";
				return NULL;
		}
	}

	/* ATM static & same on all calls! */
	buffer = menu_path(NULL);

	/* if not absolute path, perpend current work directory */
	if (idx < argc && argv[idx] && argv[idx][0] == '/') {
		strcpy(buffer, argv[idx]);
	} else {
		getcwd(buffer, MAX_PATH_LEN);
		if (idx < argc && argv[idx]) {
			len = strlen(buffer);
			buffer[len++] = '/';
			strcpy(&buffer[len], argv[idx]);
		}
	}
	/* check directory validity */
	if(!(dfd = opendir(buffer))) {
		free_path(buffer);
		*error = "directory access failed";
		return NULL;
	}
	closedir(dfd);

	/* assert '/' at the end */
	len = strlen(buffer);
	if (buffer[len-1] != '/') {
		buffer[len++] = '/';
		buffer[len] = '\0';
	}
	/* read default icons */
	strcpy(&buffer[len], DIR_ICON);
	FolderIcon = w_readpbm(buffer);
	strcpy(&buffer[len], FILE_ICON);
	FileIcon = w_readpbm(buffer);
	buffer[len] = '\0';

	/* setup menu for using directory_cb for getting the dir contents */
	menu[0].string = strdup(buffer);
	menu[0].type = MenuSub;
	menu[1].type = MenuEnd;
	menu[1].string = "";

	/* read directory and return it's menu */
	directory_cb(NULL, &menu[0]);
	return menu[0].sub;
}

int main (int argc, char *argv[])
{
	widget_t *top, *shell, *menu, *time, *quit;
	long rwd, rht, wd, ht, mode;
	int elements, orient, align;
	wt_menu_item *menustruct;
	const char *font, *error;

	menustruct = initialize(argc, argv, &elements, &orient, &align, &font, &error);
	if (!menustruct) {
		fprintf(stderr, "\n%s error: %s\n", *argv, error);
		fprintf(stderr, "%s, a W program launcher.\n", *argv);
		fprintf(stderr, "usage: %s [<options)] [<path>]\n", *argv);
		fprintf(stderr, "  -f <name>	set text font\n");
		fprintf(stderr, "  -v|-h	alignment\n");
		fprintf(stderr, "  -l|-c|-r	position\n");
		fprintf(stderr, "  -t|-e	show time, errors\n");
		fprintf(stderr, "  -q		quit button\n");
		return -1;
	}

	top = wt_init();
	mode = ShellModePopup;
	shell = wt_create(wt_shell_class, top);
	wt_setopt(shell, WT_MODE, &mode, WT_EOL);
	menu = wt_create(wt_menu_class, shell);

	wt_setopt(menu,
		WT_LIST_ADDRESS, menustruct,
		WT_DRAW_FN, draw_icon_cb,
		WT_EOL);

	/* if you'll want these after the start button on the wstart line,
	 * you'll have to add them to menu after setting WT_LIST_ADDRESS.
	 */
        if (elements & ADD_TIME) {
		mode = LabelModeWithBorder;
		time = wt_create(wt_label_class, menu);
		wt_setopt(time, WT_MODE, &mode, WT_EOL);
		timer_cb((long)time);
	}
	if (elements & ADD_QUIT) {
		quit = wt_create(wt_pushbutton_class, menu);
		wt_setopt(quit,
			WT_LABEL, "Quit",
			WT_ACTION_CB, quit_cb,
			WT_EOL);
	}
	if (font) {
		/* propagates to _all_ children */
		wt_setopt(menu, WT_FONT, font, WT_EOL);
	}

	if (orient == OrientVert) {
		mode = orient;
		ht = AlignFill;
		wt_setopt(menu, WT_ALIGNMENT, &ht,
			  WT_ORIENTATION, &mode,
			  WT_EOL);
	}
	/* position lauch bar */
	wt_minsize(shell, &wd, &ht);
	rwd = WROOT->width  - WIN_BORDER * 2;
	rht = WROOT->height - WIN_BORDER * 2;
	switch (align) {
	case AlignTop:
		if (orient == OrientVert) {
			wt_reshape(shell, 0, 0, WT_UNSPEC, rht);
			mode = AlignRight;
		} else {
			wt_reshape(shell, 0, 0, rwd, WT_UNSPEC);
			mode = AlignBottom;
		}
		wt_setopt(menu, WT_ALIGNMENT, &mode, WT_EOL);
		break;
	case AlignCenter:
		if (orient == OrientVert) {
			wt_reshape(shell, (rwd - wd) / 2, 0, WT_UNSPEC, rht);
		} else {
			wt_reshape(shell, 0, (rht - ht) / 2, rwd, WT_UNSPEC);
		}
		break;
	case AlignBottom:
		if (orient == OrientVert) {
			wt_reshape(shell, rwd - wd, 0, WT_UNSPEC, rht);
			mode = AlignLeft;
		} else {
			wt_reshape(shell, 0, rht - ht, rwd, WT_UNSPEC);
			mode = AlignTop;
		}
		wt_setopt(menu, WT_ALIGNMENT, &mode, WT_EOL);
		break;
	}
	wt_realize(top);
	wt_run();
	return 0;
}

